### ISSUE TITLE
Cheogram Support

### ISSUE THREAD

**Issue by trymeouteh** [dino/dino#995]
Please add a setting for any XMPP account to have cheogram.com support.

When this is enabled, it will grab contacts from the systems contacts list such as GNOME contacts and get the phone numbers and names. This will allow you to send a message to any number such as +12223334444@cheogram.com and the contact name will be displayed as the name of the contact from your contacts list. This will allow for uses to have a good user experience when using Dino with a jmp.chat or soprani.ca account which is a service that allows you to send and receive SMS messages over XMPP.

Just a simple checkbox setting in the account setting to enable or disable this.

**Comment by ItsRogueRen** [comment]
Bumping this, Dino is used exclusively to access my JMP.chat texts on my PC and not having contact names is a nightmare in group texts

**Comment by jgarte** [comment]
I would also like this feature.

Leaving a jmp link here for reference:

https://jmp.chat/

**Comment by waltmck** [comment]
> > Bumping this, Dino is used exclusively to access my JMP.chat texts on my PC and not having contact names is a nightmare in group texts
> 
> There are currently at least 2 open PRs that implement this: [#1212](https://github.com/dino/dino/pull/1212) and [#1282](https://github.com/dino/dino/pull/1282) . If you're running Debian 11 (stable), a package with that first patch can be found here: https://people.debian.org/~dilinger/dino-im/

For the record, I don't think that those PRs are really applicable to what this issue is requesting. They are not really Cheogram-specific and were (in my opinion, rightly) rejected because they involve overwriting message content in a non-standardized and unexpected way to work around what is really a limitation in the XMPP protocol. In my opinion, Cheogram needs to work toward standardizing the functionality they need as an extension to XMPP rather than hacking around it.

There is a much simpler and less controversial solution that would dramatically improve the experience of using Cheogram with Dino: fall back to phone numbers from system contacts to resolve @cheogram.com JID's that do not have an associated XMPP contact. Since this is a GNOME-focused project it could use GNOME Contacts' dbus API, though CalDAV would be better for portability.

**Comment by dilinger** [comment]
> 
> For the record, I don't think that those PRs are really applicable to what this issue is requesting. They are not really Cheogram-specific and were (in my opinion, rightly) rejected because they involve overwriting message content in a non-standardized and unexpected way to work around what is really a limitation in the XMPP protocol. In my opinion, Cheogram needs to work toward standardizing the functionality they need as an extension to XMPP rather than hacking around it.
> 

I agree with the assessment that we need to standardize the functionality. I mentioned to @singpolyma some time ago an idea to do it using the markup XEP (XEP-0394) by adding a roster tag, but I'm not sure it would work well with maintaining xmpp: links. Another XEP that might make more sense is the references one (XEP-0372), but that's marked experimental and I'm not aware of any clients implementing it. Eg, `<body><xmpp:+1234@cheogram.com> Hello!</body><reference xmlns='urn:xmpp:reference:0' type='roster' begin=2 end=24 />`, where a client could resolve that to a roster name and also make it a link that opens up a chat with the specific user. The references XEP already has a similar feature but in the other direction - a 'mention' type that resolves a display name to a JID. It would make sense to add a feature that changes a JID to a display name.

> There is a much simpler and less controversial solution that would dramatically improve the experience of using Cheogram with Dino: fall back to phone numbers from system contacts to resolve @cheogram.com JID's that do not have an associated XMPP contact. Since this is a GNOME-focused project it could use GNOME Contacts' dbus API, though CalDAV would be better for portability.

I don't understand your suggestion.

**Comment by waltmck** [comment]
> > For the record, I don't think that those PRs are really applicable to what this issue is requesting. They are not really Cheogram-specific and were (in my opinion, rightly) rejected because they involve overwriting message content in a non-standardized and unexpected way to work around what is really a limitation in the XMPP protocol. In my opinion, Cheogram needs to work toward standardizing the functionality they need as an extension to XMPP rather than hacking around it.
> 
> I agree with the assessment that we need to standardize the functionality. I mentioned to [@singpolyma](https://github.com/singpolyma) some time ago an idea to do it using the markup XEP (XEP-0394) by adding a roster tag, but I'm not sure it would work well with maintaining xmpp: links. Another XEP that might make more sense is the references one (XEP-0372), but that's marked experimental and I'm not aware of any clients implementing it. Eg, `<body><xmpp:+1234@cheogram.com> Hello!</body><reference xmlns='urn:xmpp:reference:0' type='roster' begin=2 end=24 />`, where a client could resolve that to a roster name and also make it a link that opens up a chat with the specific user. The references XEP already has a similar feature but in the other direction - a 'mention' type that resolves a display name to a JID. It would make sense to add a feature that changes a JID to a display name.

It's good to see that some thought is being given to this!

> > There is a much simpler and less controversial solution that would dramatically improve the experience of using Cheogram with Dino: fall back to phone numbers from system contacts to resolve @cheogram.com JID's that do not have an associated XMPP contact. Since this is a GNOME-focused project it could use GNOME Contacts' dbus API, though CalDAV would be better for portability.
> 
> I don't understand your suggestion.

My suggestion is the same as (my interpretation of) the original issue. I will try to motivate the idea more concretely.

Like most people I use a CardDAV server to organize my contacts. When I meet a new person, I open the Contacts app on my phone (or computer) and enter their name, phone number, email, address, photo, etc. The CardDAV server syncs this contact information across my devices.

Say my friend John Smith texts me from phone number `+11234567890`. In Cheogram on Android, the notification will say "John Smith" and the chat will be titled "John Smith". This is because Cheogram on Android, upon getting a message from JID `+11234567890@cheogram.com`, will take the following steps:
1. Check whether `+11234567890` is in the XMPP roster. If so, return the name of the contact in the roster. Otherwise continue to (2).
2. Check my system contacts for any contact with the phone number `+11234567890`. If one is found, display the name of that contact. Otherwise, continue to (3).
3. Display the raw JID.

In contrast, Dino will just display the raw JID `+11234567890@cheogram.com`. This is because Dino will take the following steps:
1. Check whether `+11234567890` is in the XMPP roster. If so, return the name of the contact in the roster. Otherwise continue to (2).
2. Display the raw JID.

In order to see who is texting me, I will have to manually copy and paste that number into my Contacts app to see which contact it corresponds to and add that person to my XMPP roster. There are some other [ugly hacks](https://gitlab.com/djsumdog/dav-xmpp-sync) that have arisen to work around this problem, but there is currently no reliable way to keep one's XMPP roster in sync with their CardDAV server.

The solution proposed by the OP is to match the behavior of the Cheogram on Android by attempting to use local system contacts to resolve JIDs to contact names.

Does this clear things up?
